### PR TITLE
Slash commands

### PR DESCRIPTION
**Pull request by promise**

### CODE REVIEW

File: src/handlers/commands.js [hunk 1]
```diff
@@ -1,21 +1,18 @@
...
+const commands = new Map(), aliases = new Map(), statics = require("../commands/_static.json");
+fs.readdir("./src/commands/", (err, files) => {
+  if (err) return console.log(err);
+  for (const file of files) if (file.endsWith(".js")) {
+    const commandFile = require(`../commands/${file}`), fileName = file.replace(".js", "");
+    if (config.isPremium || !commandFile.premiumOnly) {
+      commands.set(fileName, commandFile);
+      if (commandFile.aliases) for (const alias of commandFile.aliases) aliases.set(alias, fileName);
     }
-    resolve();
```
Comment (fabianneufang) [inline src/handlers/commands.js]: Resolve is not defined here because it is not a promise

File: src/handlers/commands.js [hunk 2]
```diff
@@ -67,21 +64,19 @@ module.exports.setupSlashCommands = async client => {
...
   await Promise.all(slashCommands
     .filter(c => !commands.get(slashCommand.name))
     .map(({ id }) => 
-      client.api.applications(config.id).commands.delete(id)
+      client.api.applications(client.user.id).commands.delete(id)
     )
   )
 
   // register commands
   await Promise.all(commands.map(async ({ description, options, permissionRequired }, name) => {
```
Comment (fabianneufang) [inline src/handlers/commands.js]: Commands is a map and therefore not mappable

**Review by fabianneufang — CHANGES_REQUESTED** [review]
The code will not run because

File: src/handlers/commands.js [hunk 3]
```diff
@@ -67,21 +64,19 @@ module.exports.setupSlashCommands = async client => {
...
 
-module.exports.registerSlashCommands = async () => {
-  await commandLoader;
-
+module.exports.registerSlashCommands = async client => {
   // remove old commands
-  const slashCommands = await client.api.applications(config.id).commands.get();
+  const slashCommands = await client.api.applications(client.user.id).commands.get();
   await Promise.all(slashCommands
     .filter(c => !commands.get(slashCommand.name))
```
Comment (fabianneufang) [inline src/handlers/commands.js]: `slashCommand` is not defined

**Review by fabianneufang — CHANGES_REQUESTED** [review]

File: src/handlers/commands.js (hunk 2 above)
Comment (promise) [reply src/handlers/commands.js]: Not mapping the commands variable, mapping slashCommands which is a list and not a Map/Collection.

**Comment by promise** [comment]
Fabi, it's a draft for a reason 😤

File: src/handlers/commands.js (hunk 2 above)
Comment (fabianneufang) [reply src/handlers/commands.js]: I added the comment on line L83 I don't know why GitHub displays that abvove

**Comment by aditya20-b** [comment]
staph talking in here 😤

**Comment by fabianneufang** [comment]
Just wanted to review it

**Comment by promise** [comment]
I'll request reviews when I want them on drafts lol

**Comment by fabianneufang** [comment]
This was just meant as a help 😥

File: src/handlers/commands.js (hunk 1 above)
Comment (promise) [reply src/handlers/commands.js]: Fixed in https://github.com/countr/countr/pull/83/commits/c6b4fb30f053d84a94eed6278523f4ccb4e86b43

File: src/handlers/commands.js (hunk 2 above)
Comment (promise) [reply src/handlers/commands.js]: Fixed in https://github.com/countr/countr/pull/83/commits/c6b4fb30f053d84a94eed6278523f4ccb4e86b43

File: src/handlers/commands.js (hunk 3 above)
Comment (promise) [reply src/handlers/commands.js]: Fixed in https://github.com/countr/countr/pull/83/commits/c6b4fb30f053d84a94eed6278523f4ccb4e86b43

**Comment by promise** [comment]
no worries

File: src/handlers/commands.js [hunk 4]
```diff
@@ -46,36 +46,101 @@ module.exports = async (message, gdb, db, countingChannel, prefix) => {
   if (message.channel.id == countingChannel) setTimeout(() => message.channel.id == gdb.get().channel ? deleteMessages([ message, response ]) : null, 5000);
 };
 
-module.exports.setupSlashCommands = async client => {
+module.exports.setupSlashCommands = async (client, db) => {
   client.ws.on("INTERACTION_CREATE", async interaction => {
     const commandFile = commands.get(interaciton.data.name);
```
Comment (fabianneufang) [inline src/handlers/commands.js]: interaciton is not defined

**Review by fabianneufang — CHANGES_REQUESTED** [review]
Typo in interaction.

File: src/handlers/commands.js (hunk 4 above)
Comment (fabianneufang) [reply src/handlers/commands.js]: So here: L51 `interaciton`

File: src/handlers/commands.js [hunk 5]
```diff
@@ -46,36 +46,101 @@ module.exports = async (message, gdb, db, countingChannel, prefix) => {
...
     if (!commandFile) return respondWithError(interaction, `❌ The command \`${interaction.data.name}\` does not exist.`)
 
     const guild = client.guilds.cache.get(interaction_guild_id);
```
Comment (fabianneufang) [inline src/handlers/commands.js]: Change to `interaction.guild_id`

File: src/handlers/commands.js [hunk 6]
```diff
@@ -46,36 +46,101 @@ module.exports = async (message, gdb, db, countingChannel, prefix) => {
...
+        acknowledge: () => client.api.interactions(interaction.id, interaction.token).callback.post({ data: { type: 5 } }),
+      },
+      { client, guild, channel, member },
+      args, gdb,
+      { permissionLevel, db }
+    )
   })
 }
 
 const respondWithError = ({ id, token }, content) => client.api.interactions(id, token).callback.post({ data: { type: 3, data: { flags: 64, content }}})
```
Comment (fabianneufang) [inline src/handlers/commands.js]: `client` is not defined

**Review by fabianneufang — CHANGES_REQUESTED** [review]

File: src/handlers/commands.js [hunk 7]
```diff
@@ -46,36 +46,101 @@ module.exports = async (message, gdb, db, countingChannel, prefix) => {
...
+module.exports.setupSlashCommands = async (client, db) => {
   client.ws.on("INTERACTION_CREATE", async interaction => {
     const commandFile = commands.get(interaciton.data.name);
     if (!commandFile) return respondWithError(interaction, `❌ The command \`${interaction.data.name}\` does not exist.`)
 
     const guild = client.guilds.cache.get(interaction_guild_id);
     if (!guild) return respondWithError(interaction, `❌ This guild is currently unavailable.`)
     const channel = guild.channels.cache.get(interaction.channel_id);
-    if (!channel) return respondWithError(interaction, `❌ I don't have access to view this channel.`)
+    if (!channel || !(commandFile.needAccessToChannel && !channel.viewable)) return respondWithError(interaction, `❌ I don't have access to view this channel.`)
```
Comment (fabianneufang) [inline src/handlers/commands.js]: Ln 57 remove negation of `(commandFile.needAccessToChannel && !channel.viewable)`

File: src/handlers/commands.js [hunk 8]
```diff
@@ -46,36 +46,101 @@ module.exports = async (message, gdb, db, countingChannel, prefix) => {
...
+
+    const gdb = await db.guild(guild.id), { channel: countingChannel } = gdb.get();
+    if (channel.id == countingChannel && !commandFile.allowInCountingChannel) return respondWithError(interaction, "❌ This command is disabled inside the counting channel.")
+
     const member = guild.members.fetch(interaction.member.user.id);
```
Comment (fabianneufang) [inline src/handlers/commands.js]: await the promise

**Review by fabianneufang — CHANGES_REQUESTED** [review]

File: src/handlers/commands.js (hunk 6 above)
Comment (fabianneufang) [reply src/handlers/commands.js]: Possible fix:
`client => ({ id, token }, content) => client.api.interactions(id, token).callback.post({ data: { type: 3, data: { flags: 64, content }}})`

and then set the `respondWithError` above in the `INTERACTION_CREATE` callback like `const respondWithError = respondWithErrorCallback(client)`

**Comment by promise** [comment]
I'll ask for a review next time, please stop reviewing. It's a draft for a reason.

**Comment by fabianneufang** [comment]
Ok I won't review it anymore.

File: src/handlers/slashCommands.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
+const fs = require("fs"), config = require("../../config.json");
+
+module.exports = async (client, db, shardid) => {
+  if (shardid == 0) registerCommands(client).then(() => console.log("Manager: Slash Commands have been registered."));
+
+  client.ws.on("INTERACTION_CREATE", async interaction => {
+    const
+      commandFile = require(`../commands/slash/${interaction.data.name}.js`),
+      gdb = await db.guild(interaction.guild_id),
```
Comment (fabianneufang) [inline src/handlers/slashCommands.js:10]: This could fail when the command is executed in dms

**Review by fabianneufang — CHANGES_REQUESTED** [review]
Everything fine except for dms

**Review by fabianneufang — APPROVED** [review]

**Review by lacuna-exe — CHANGES_REQUESTED** [review]
As ephemeral messages don't support embeds using the scoreboard slash command doesn't work. [Image](https://i.e5510.workers.dev/bC1JfD3.png)

**Review by fabianneufang — APPROVED** [review]

**Review by aditya20-b — APPROVED** [review]

**Comment by promise** [comment]
> As ephemeral messages don't support embeds using the scoreboard slash command doesn't work. [Image](https://i.e5510.workers.dev/bC1JfD3.png)

image

pushing a fix in a minute

**Comment by fabianneufang** [comment]
> > As ephemeral messages don't support embeds using the scoreboard slash command doesn't work. [Image](https://i.e5510.workers.dev/bC1JfD3.png)
> 
> image
> 
> pushing a fix in a minute

Could we generate an image server side and send that to the user?

**Comment by promise** [comment]
> Could we generate an image server side and send that to the user?

we would have to fetch all the users first, and then display their tag or nickname and stuff. Would be a hassle. Can just send it without the embed instead.

**Comment by fabianneufang** [comment]
> we would have to fetch all the users first, and then display their tag or nickname and stuff. Would be a hassle. Can just send it without the embed instead.

thought about caching or something but that would be a hassle to handle

**Review by fabianneufang — APPROVED** [review]

**Review by lacuna-exe — APPROVED** [review]

**Pull request merged**